gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/JDBCTokenStateService.java

```diff
@@ -0,0 +1,180 @@
+      }
+    } catch (SQLException e) {
+      log.errorRemovingTokensFromDatabase(tokenIds.size(), e.getMessage(), e);
+    }
+    super.removeTokens(tokenIds);
+  }
+
+  @Override
+  public long getTokenExpiration(String tokenId, boolean validate) throws UnknownTokenException {
+    try {
+      // check the in-memory cache, then
+      return super.getTokenExpiration(tokenId, validate);
+    } catch (UnknownTokenException e) {
+      // It's not in memory
+    }
+
+    long expiration = 0;
+    try {
+      expiration = tokenDatabase.getTokenExpiration(tokenId);
+      log.fetchedExpirationFromDatabase(Tokens.getTokenIDDisplayText(tokenId), expiration);
+    } catch (SQLException e) {
+      log.errorFetchingExpirationFromDatabase(Tokens.getTokenIDDisplayText(tokenId), e.getMessage(), e);
+    }
+    return expiration;
+  }
+
+  @Override
+  protected void updateExpiration(String tokenId, long expiration) {
+    // Update in-memory
+    super.updateExpiration(tokenId, expiration);
+
+    try {
+      final boolean updated = tokenDatabase.updateExpiration(tokenId, expiration);
+      if (updated) {
+        log.updatedExpirationInDatabase(Tokens.getTokenIDDisplayText(tokenId), expiration);
+      } else {
+        log.failedToUpdateExpirationInDatabase(Tokens.getTokenIDDisplayText(tokenId), expiration);
+      }
+    } catch (SQLException e) {
+      log.errorUpdatingExpirationInDatabase(Tokens.getTokenIDDisplayText(tokenId), e.getMessage(), e);
+    }
+  }
+
+  @Override
+  protected long getMaxLifetime(String tokenId) {
+    long maxLifetime = super.getMaxLifetime(tokenId);
+
+    // If there is no result from the in-memory collection, proceed to check the Database
+    if (maxLifetime < 1L) {
+      try {
+        maxLifetime = tokenDatabase.getMaxLifetime(tokenId);
+        log.fetchedMaxLifetimeFromDatabase(Tokens.getTokenIDDisplayText(tokenId), maxLifetime);
+      } catch (SQLException e) {
+        log.errorFetchingMaxLifetimeFromDatabase(Tokens.getTokenIDDisplayText(tokenId), e.getMessage(), e);
+      }
+    }
+    return maxLifetime;
+  }
+
+  @Override
+  protected boolean isUnknown(String tokenId) {
+    boolean isUnknown = super.isUnknown(tokenId);
+
+    // If it's not in the cache, then check in the Database
+    if (isUnknown) {
+      try {
+        isUnknown = tokenDatabase.getMaxLifetime(tokenId) < 0;
+      } catch (SQLException e) {
+        log.errorFetchingMaxLifetimeFromDatabase(Tokens.getTokenIDDisplayText(tokenId), e.getMessage(), e);
+      }
+    }
+    return isUnknown;
+  }
+
+  @Override
+  protected List<String> getTokenIds() {
+    List<String> tokenIds = new LinkedList<>();
+    try {
+      tokenIds = tokenDatabase.getTokenIds();
+      log.fetchedAllTokenIdsFromDatabase(tokenIds.size());
+    } catch (SQLException e) {
+      log.errorFetchingAllTokenIdsFromDatabase(e.getMessage(), e);
+    }
+    return tokenIds;
+  }
+
+  @Override
+  public void addMetadata(String tokenId, TokenMetadata metadata) {
```

Comment:
It seems that the addMetadata is decoupled from the addToken. I can imagine this came about due to the alias service based approach but it seems sort of artificial for JDBC approach. Perhaps this can be refactored in the future to deprecate this method and include an addToken that accepts a TokenMetadata object as well? Should we actually do that now and leave the separate method for where it is currently used but deprecate it?

Reply:
Good idea; let me submit a new PS with that change.

Reply:
@lmccay I went through the code and checked the current usage of these methods and analyzed the amount of changes to be made and I decided to handle this as a separate JIRA: https://issues.apache.org/jira/browse/KNOX-2580